mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/eventsourcing/RegisteredGroupsAggregate.java

```diff
@@ -0,0 +1,165 @@
+                "Trying to remove a non existing group");
+
+            return new State(ImmutableMap.<Group, Status>builder()
+                .putAll(event.getAddedGroups()
+                    .stream()
+                    .collect(Guavate.toImmutableMap(Functions.identity(), any -> Status.USED)))
+                .putAll(notIn(Sets.union(event.getAddedGroups(), event.getRemovedGroups())))
+                .putAll(event.getRemovedGroups()
+                    .stream()
+                    .collect(Guavate.toImmutableMap(Functions.identity(), any -> Status.UNUSED_BUT_BINDED)))
+                .build());
+        }
+
+        private State apply(UnbindSucceededEvent event) {
+            Preconditions.checkArgument(bindedGroups().contains(event.getGroup()), "unbing a non binded group," +
+                " or a used group");
+
+            return new State(ImmutableMap.<Group, Status>builder()
+                .putAll(notIn(ImmutableSet.of(event.getGroup())))
+                .build());
+        }
+
+        private ImmutableSet<Group> usedGroups() {
+            return groups.entrySet()
+                    .stream()
+                    .filter(group -> group.getValue() == Status.USED)
+                    .map(Map.Entry::getKey)
+                    .collect(Guavate.toImmutableSet());
+        }
+
+        private ImmutableSet<Group> bindedGroups() {
+            return groups.entrySet()
+                    .stream()
+                    .filter(group -> group.getValue() == Status.UNUSED_BUT_BINDED)
+                    .map(Map.Entry::getKey)
+                    .collect(Guavate.toImmutableSet());
+        }
+
+        private ImmutableMap<Group, Status> notIn(Set<Group> toBeFilteredOut) {
+            return groups.entrySet()
+                    .stream()
+                    .filter(group -> !toBeFilteredOut.contains(group.getKey()))
+                    .collect(Guavate.entriesToImmutableMap());
+        }
+    }
+
+    public static RegisteredGroupsAggregate load(History history) {
+        return new RegisteredGroupsAggregate(history);
+    }
+
+    private final History history;
+    private State state;
+
+    private RegisteredGroupsAggregate(History history) {
+        this.history = history;
+        this.state = State.initial();
+
+        history.getEventsJava()
+            .forEach(this::apply);
+    }
+
+    public List<RegisteredGroupListenerChangeEvent> handle(RequireGroupsCommand requireGroupsCommand, Clock clock) {
+        List<RegisteredGroupListenerChangeEvent> detectedChanges = detectChanges(requireGroupsCommand, clock);
+        detectedChanges.forEach(this::apply);
+        return detectedChanges;
+    }
+
+    public List<UnbindSucceededEvent> handle(MarkUnbindAsSucceededCommand command) {
+        UnbindSucceededEvent event = new UnbindSucceededEvent(history.getNextEventId(), command.getSucceededGroup());
```

Comment:
it should fail if groups are not tracked by the state.
Also, it should have the eventId because we may have recreate the group.